### ISSUE TITLE
Enforce xsnap version on chain start

### ISSUE THREAD

**Issue by mhofman** [Agoric/Agoric#7012]
## What is the Problem Being Solved?

If an upgrade contains an xsnap-worker minor version change (the kind that is snapshot and JS execution compatible), the validator must build and use this new version. See https://github.com/Agoric/agoric-sdk/issues/6361

If they don't their node will likely fall off consensus at some point. While our update instructions include `yarn install && yarn build`, which will rebuild XS, the chain process would likely not fail early enough if those steps were not performed, resulting in a hard to recover state.

## Description of the Design

Include a check in when swingset or cosmic-swingset starts that the xsnap version is the expected one.

This check should be by-passable so that we can manually use other XS revisions that are supposed to be consensus matching.

Likely related to https://github.com/Agoric/agoric-sdk/issues/6596

## Security Considerations

This would ensure all nodes will run as expected after an upgrade, guaranteeing the health of the chain.

**Comment by warner** [comment]
@mhofman What were you thinking we could use as a version number/string? If the goal is to detect when agoric-sdk has been updated but `packages/xsnap` has not been recompiled, then it sounds like we need to put something in the compiled binary, which can either be reported with e.g. `xsnap --version`, or returned as part of some special netstring-pipe command. And then compared against.. something we check into the `packages/xsnap/` JS code?

**Comment by warner** [comment]
@ivanlei and I figure that it'd be nice to have this, but it's not such a huge footgun that we really need it in this release. Validators have multiple ways to fail to recompile everything, but failing to run `yarn build` is not as big as failing to run `make` (to rebuild the golang code). So we're kicking this out of the release.

**Comment by mhofman** [comment]
Given that we have stricter consensus checks now (snapshot hashes, and soon computrons per delivery), and that a validator can recover by using state-sync, I agree this is not really needed anymore.

**Comment by warner** [comment]
Note to self: `xsCommon.h` has a triplet of version constants:

https://github.com/agoric-labs/moddable/blob/428864a754c48486f42bd43a1fdf3e3de8aaefe6/xs/sources/xsCommon.h#L107-L109

* 1: these would only represent changes to the upstream XS code, on release boundaries
  * not any changes we might cherry-pick from between releases
  * and not any local changes we apply that don't go into upstream
* 2: we'd need to export them somehow
  * probably return them from the `setBundle` command, or some other command that happens very early, and only once per worker

**Comment by warner** [comment]
@ivanlei and @arirubinstein and I added a hack just now, in #7831 and https://github.com/agoric-labs/xsnap-pub/pull/41 , to add an `xsnap -n` command which emits `agoric-upgrade-10` (the "network name" of the new chain release, so `-n` stands for "network upgrade version"), and to have our `bin/agd` shell script use a relative path (relative to `thisdir`, a one-level -deep `readline` of the `agd` being invoked), to call `xsnap -n` and insist that output is `agoric-upgrade-10`.

This is enough to solve today's problem, which is to insist that xsnap has been recompiled recently (at least after the landing of #7831). We'll need to reconsider this more carefully next week, to solve the next problem, which begins when we make any change to XS or xsnap, or when we want to introduce the `agoric-upgrade-11` handler into the cosmos/golang code (to facilitate automated upgrade-from-current-mainnet tests, and which really ought to come with a change to `agd` to look for `agoric-upgrade-11` instead of `-10`). Some things we'd like to improve:

* `xsnap -v` already includes the XS version triple, and a string copied in from the package.json version of `packages/xsnap` (aka `@agoric/xsnap`). This is pretty close to what we want, so it'd be nice to use it directly somehow.
  * like, maybe we could update `agd` to expect `xsnap -v` to match a particular value
  * but, our build/release process (MAINTAINERS.md) updates the package.json versions in one fell swoop, and we'd need to modify `bin/agd` *after* running that process and learning the new version, but then does `bin/agd` get modified in the same commit as the one `lerna version` creates? Or do we have to add an extra commit, which would mean the sdk-wide `agoric-sdk` tag wouldn't include the newer `agd`.
  * and, if we make changes to `xsnap-platform.c`, they wouldn't be included in the `xsnap -v` output until after we update `packages/xsnap/` and perform a version-bumping release sequence, whereas I'd really prefer to discover the mismatch earlier than that
* `agd` now has detailed knowledge about where the xsnap executable lives, which I'd prefer to be hidden within `packages/xsnap`
  * if `agd` weren't a shell script, we could add a (JS) API to `packages/xsnap` which runs `xsnap -v` and reports back the string
  * maybe `packages/xsnap` could have a tool at a well-known path, which `agd` could invoke with `node .../packages/xsnap/tools/xsnap-version.js`
  * but really, `packages/xsnap/src/xsnap.js` should be the outer half, not `agd`
* frequency of update: ideally we'd like this string to be updated upon any change to `XS` or `xsnap`. Incrementing a counter on both sides is simple but manual and prone to being forgotten. In a perfect world this would be driven by git hashes somehow.
